physionet-django/training/views.py

```diff
@@ -1,31 +1,87 @@
-from calendar import c
-from hmac import new
 import json
-import operator
-from itertools import chain
-import re
 
 from django.contrib import messages
 from django.contrib.auth.decorators import login_required, permission_required
-from django.db import transaction
-from django.db.models import Prefetch
 from django.http import JsonResponse
 from django.shortcuts import get_object_or_404, redirect, render
 from django.utils import timezone
 from django.utils.crypto import get_random_string
+from django.http import Http404
+from django.urls import reverse
+
 from project.validators import validate_version, is_version_greater
 
 from rest_framework.parsers import JSONParser
 
-from user.models import Training, TrainingType, TrainingQuestion, RequiredField
+from user.models import Training, TrainingType, RequiredField
 from user.enums import TrainingStatus
 
-from training.models import Course, Quiz, QuizChoice, ContentBlock
+from training.models import Course, Quiz, QuizChoice, ContentBlock, Module
 from training.models import CourseProgress, ModuleProgress, CompletedContent, CompletedQuiz
 from training.serializers import CourseSerializer
 from console.views import console_permission_required
 
 
+# Defining few utility functions to help with the views
+def get_course_and_module(training_slug, module_order):
+    course = Course.objects.filter(training_type__slug=training_slug, is_active=True).order_by('version').last()
+    if course is None:
+        raise Http404()
+    module = get_object_or_404(course.modules, order=module_order)
+    return course, module
+
+
+def get_course_and_module_progress(user, course, module_order):
+    course_progress, _ = CourseProgress.objects.get_or_create(user=user, course=course)
+    module = get_object_or_404(course.modules, order=module_order)
+    module_progress, _ = ModuleProgress.objects.get_or_create(course_progress=course_progress, module=module)
+    return course_progress, module_progress
+
+
+def handle_quiz_submission(module_progress, quiz, choice_id):
+    correct_choice = QuizChoice.objects.filter(quiz=quiz, is_correct=True).first()
+    if correct_choice.id != int(choice_id):
+        return False
+    return True
+
+
+def update_module_progress(module_progress, order):
+    if order > module_progress.last_completed_order:
+        module_progress.last_completed_order = order
+        module_progress.save()
+
+
+def get_min_max_order_in_module(module):
+    # calculating the max and min order of the module
+    module_blocks = ContentBlock.objects.filter(module=module).values_list('order', flat=True)
+    quiz_blocks = Quiz.objects.filter(module=module).values_list('order', flat=True)
+    module_max_order = max(max(module_blocks, default=0), max(quiz_blocks, default=0))
+    module_min_order = min(min(module_blocks, default=0), min(quiz_blocks, default=0))
+    return module_min_order, module_max_order
+
+
+def handle_course_completion(course, course_progress):
```

Comment:
Please could you add docstrings for these functions. Their purpose isn't always clear (e.g. "handle course completion" does not give much information about the purpose of this function).